scripts/update_todo_references.py

```diff
@@ -0,0 +1,248 @@
+
+        regex = re.compile(
+            r"^[ \t]*# TODO ?\[("
+            + "|".join(todos_to_delete)
+            + r")\] ?:(.*$\n(^[ \t]*# {2}.*$\n)*)",
+            flags=re.MULTILINE
+        )
+        new_file_content = re.sub(
+            regex,
+            "",
+            file_content,
+        )
+
+        file.write_text(new_file_content)
+        print(f"Deleted {len(todos_to_delete)} todos in file '{file}'.")
+
+
+def find_todos(file: Path) -> Tuple[List[Todo], List[Todo]]:
+    """Finds referenced and unreferenced Todos in a file.
+
+    Args:
+        file: The file to search.
+
+    Returns: A tuple (unreferenced_todos, referenced_todos) for todos
+        without/with reference to jira issues.
+    """
+    file_content = file.read_text()
+
+    matches = re.findall(
+        r"^[ \t]*# TODO ?\[(LOW|MEDIUM|HIGH|[A-Z]*?-[0-9]*?)\] ?:"
+        r"(.*$\n(^[ \t]*# {2}.*$\n)*)",
+        file_content,
+        flags=re.MULTILINE,
+    )
+
+    unreferenced_todos = []
+    referenced_todos = []
+
+    for priority_or_key, description, _ in matches:
+        # remove whitespace and leading '#' from description
+        lines = description.split("\n")
+        lines = [line.strip().lstrip("#").strip() for line in lines]
+        description = " ".join(lines)
+
+        todo = Todo(filepath=str(file), description=description)
+        if priority_or_key in PRIORITIES:
+            todo.priority = priority_or_key
+            unreferenced_todos.append(todo)
+        else:
+            todo.issue_key = priority_or_key
+            referenced_todos.append(todo)
+
+    print(f"Found {len(unreferenced_todos)} unreferenced and "
+          f"{len(referenced_todos)} referenced todos in file '{file}'.")
+
+    return unreferenced_todos, referenced_todos
+
+
+def update_file_with_issue_keys(file: Path, todos: List[Todo]) -> None:
+    """Replaces TODO priorities by the issue key that the TODO is referencing.
+
+    Args:
+        file: The file in which the todos will be replaced.
+        todos: The todos to replace.
+    """
+    file_content = file.read_text()
+    todo_iterator = iter(todos)
+
+    def _replace(match):
+        todo = next(todo_iterator)
+        result = match.group(1) + f"{todo.issue_key}" + match.group(3)
+        return result
+
+    new_file_content = re.sub(
+        r"(^[ \t]*# TODO ?\[)(LOW|MEDIUM|HIGH)(\] ?:"
+        r".*$\n(^[ \t]*# {2}.*$\n)*)",
+        _replace,
+        file_content,
+        flags=re.MULTILINE,
+    )
+
+    file.write_text(new_file_content)
+    print(f"Added issue references for {len(todos)} todos in file '{file}'.")
+
+
+def update_todo_references() -> None:
+    """Updates TODO comment references.
+
+    This method
+    - creates JIRA issues for all TODO comments that don't reference
+        an issue yet and
+    - removes TODO comments that reference a JIRA issue that was closed
+    """
+    zenml_root = Path("src/zenml")
+    # test_root = Path("tests")
```

Comment:
Take this out?

Reply:
That's the question, do we want it to work for todos in our tests as well? There are quite a few in there, what do you think?

Reply:
I would say we keep it in

Reply:
I made it configurable via CLI argument and only run it on `src` for now, if we need the additional `tests` TODOs in the future adding them should just be copying the step of the GH Action workflow